### PR TITLE
Process stats with pid fix

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

based on https://github.com/kubernetes/kubernetes/pull/115216 by @bobbypage   and https://github.com/kubernetes/kubernetes/pull/123243 by @kannon92 

#123243 the third commit will fix the containerd cadvisor summary e2e in `pull-kubernetes-node-e2e-containerd`.

#### Which issue(s) this PR fixes:
xref https://github.com/kubernetes/kubernetes/issues/115215 https://github.com/kubernetes/kubernetes/issues/107063

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by kannon92** [comment]
Interesting! your containerd one worked!

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 1]
```diff
@@ -584,16 +585,11 @@ func (p *criStatsProvider) addSwapStats(
...
-	cs *statsapi.ContainerStats,
+	container *cadvisorapiv2.ContainerInfo,
 ) {
-	// try get process stats from cadvisor only.
-	info := getCadvisorPodInfoFromPodUID(podUID, allInfos)
-	if info != nil {
-		ps.ProcessStats = cadvisorInfoToProcessStats(info)
-		return
-	}
+	processStats := cadvisorInfoToProcessStats(container)
```
Comment (kannon92) [inline pkg/kubelet/stats/cri_stats_provider.go:590]: can you add this to cadvisor_stats_provider?

File: pkg/kubelet/eviction/helpers.go [hunk 2]
```diff
@@ -739,6 +739,10 @@ func process(stats statsFunc) cmpFunc {
 		p1Process := processUsage(p1Stats.ProcessStats)
 		p2Process := processUsage(p2Stats.ProcessStats)
 		// prioritize evicting the pod which has the larger consumption of process
+		if p1Process == 0 || p2Process == 0 {
```
Comment (kannon92) [inline pkg/kubelet/eviction/helpers.go:742]: Once you figure out the issue, I think we could remove this?

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 1 above)
Comment (pacoxu) [reply pkg/kubelet/stats/cri_stats_provider.go:590]: Good idea. Updated.

Comment (pacoxu) [reply pkg/kubelet/stats/cri_stats_provider.go:590]: With https://github.com/kubernetes/kubernetes/pull/123369/commits/baa7e4b44d3905a374f5b0409ce915d85836c0f0, the crio e2e still failed.

**Comment by kannon92** [comment]
/test pull-kubernetes-node-swap-fedora
trying cgroupv2 crio test

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 1 above)
Comment (pacoxu) [reply pkg/kubelet/stats/cri_stats_provider.go:590]: Do we have to fix the that in this PR as well? This at least will fix the containerd CI.

Comment (kannon92) [reply pkg/kubelet/stats/cri_stats_provider.go:590]: Yea, feel free to drop the fix in cadvisor stats provider. I’ll open it up and investigate it on my end.

File: pkg/kubelet/eviction/helpers.go [hunk 3]
```diff
@@ -739,6 +739,10 @@ func process(stats statsFunc) cmpFunc {
...
+			// Looks like there is an issue where cadvisor stats reports 0 processes
+			klog.V(4).InfoS("ProcessStats", "p1Process", p1Process, "p2Process", p2Process)
```
Comment (haircommander) [inline pkg/kubelet/eviction/helpers.go:744]: can we make this a more informative message if we're leaving it?

Comment (bobbypage) [reply pkg/kubelet/eviction/helpers.go:744]: this sounds like a debugging message, maybe worth to drop it?

Comment (pacoxu) [reply pkg/kubelet/eviction/helpers.go]: > this sounds like a debugging message, maybe worth to drop it?

I dropped it.

File: test/e2e_node/summary_test.go [hunk 4]
```diff
@@ -259,7 +259,7 @@ var _ = SIGDescribe("Summary API", framework.WithNodeConformance(), func() {
 					"InodesUsed":     bounded(0, 1e8),
 				}),
 				"ProcessStats": ptrMatchAllFields(gstruct.Fields{
-					"ProcessCount": bounded(0, 1e8),
+					"ProcessCount": bounded(1, 1e8),
```
[the remark this answers is not in the document]
Comment (pacoxu) [reply test/e2e_node/summary_test.go]: Done.

**Comment by kannon92** [comment]
/triage accepted
/priority important-soon

/lgtm

/assign @mrunalp @SergeyKanzhelev 

IMO even if we don't have a fix for crio yet, it may be worth considering cherry-picking this change as I think it impacts all supported versions of Kubernetes.

**Comment by pacoxu** [comment]
/hold 
for https://github.com/kubernetes/kubernetes/pull/124101 will fix both cri-o and containerd

**Comment by pacoxu** [comment]
/close 
as https://github.com/kubernetes/kubernetes/pull/124101

**Pull request closed without merging**